File: cpp/utility.cc
/*
 *
 * Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

#include "utility.h"

#include <cassert>
#include <iostream>

#include "absl/memory/memory.h"
#include "grpcpp/ext/admin_services.h"
#include "grpcpp/xds_server_builder.h"

namespace traffic_director_grpc_examples {

const char* kArgCreds = "--creds";
const char* kInsecureCreds = "insecure";
const char* kXdsCreds = "xds";

const char* ParseCommandLineArgForCredsType(int argc, char** argv) {
  for (int i = 1; i < argc; ++i) {
    std::string arg = argv[i];
    size_t start_pos = arg.find(kArgCreds);
    if (start_pos != std::string::npos) {
      start_pos += strlen(kArgCreds);
      if (arg[start_pos] == '=') {
        absl::string_view arg_val = arg.substr(start_pos + 1);
        if (arg_val == kXdsCreds) {
          return kXdsCreds;
        } else if (arg_val == kInsecureCreds) {
          return kInsecureCreds;
        } else {
          std::cerr << "Allowed values for --creds are \'xds\', \'insecure\'";
          exit(1);
        }
      } else {
        std::cerr << "The only correct argument syntax is --creds=<value>";
        exit(1);
      }
    }
  }
  return kInsecureCreds;
}

std::shared_ptr<grpc::ChannelCredentials> GetChannelCredetials(
    absl::string_view creds_type) {
  if (creds_type == kInsecureCreds) {
    return grpc::InsecureChannelCredentials();
  } else if (creds_type == kXdsCreds) {
    return grpc::experimental::XdsCredentials(
        grpc::InsecureChannelCredentials());
  }
  assert(0);
}

std::unique_ptr<grpc::ServerBuilder> GetServerBuilder(
    absl::string_view creds_type) {
  if (creds_type == kInsecureCreds) {
    return absl::make_unique<grpc::ServerBuilder>();
  } else if (creds_type == kXdsCreds) {
    return absl::make_unique<grpc::experimental::XdsServerBuilder>();
  }
  assert(0);
}

std::shared_ptr<grpc::ServerCredentials> GetServerCredentials(
    absl::string_view creds_type) {
  if (creds_type == kInsecureCreds) {
    return grpc::InsecureServerCredentials();
  } else if (creds_type == kXdsCreds) {
    return grpc::experimental::XdsServerCredentials(
        grpc::InsecureServerCredentials());
  }
  assert(0);
}

std::unique_ptr<grpc::Server> StartAdminServer(const std::string& port) {
  std::string server_address = "localhost" + port;
  grpc::EnableDefaultHealthCheckService(true);
  grpc::ServerBuilder builder;
  builder.AddListeningPort(server_address, grpc::InsecureServerCredentials());
  grpc::AddAdminServices(&builder);
  std::cout << "Admin and Health Check Server listening on " << server_address
            << std::endl;
  return builder.BuildAndStart();
}

}  // namespace traffic_director_grpc_examples

File: go/utility/utility.go
/*
 *
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

package utility

import (
	"context"
	"fmt"
	"log"
	"math/rand"
	"net"
	"os"
	"strconv"
	"time"

	"google.golang.org/grpc"
	"google.golang.org/grpc/admin"
	"google.golang.org/grpc/codes"
	accountpb "google.golang.org/grpc/grpc-wallet/grpc/examples/wallet/account"
	"google.golang.org/grpc/health"
	healthpb "google.golang.org/grpc/health/grpc_health_v1"
	"google.golang.org/grpc/metadata"
	"google.golang.org/grpc/reflection"
	"google.golang.org/grpc/status"
)

func init() {
	rand.Seed(time.Now().UnixNano())
}

// PrintHostname pulls the hostname from a response header and prints it.
func PrintHostname(header metadata.MD) {
	hostnames, ok := header["hostname"]
	if !ok {
		log.Printf("server host: error: no hostname")
		return
	}
	if len(hostnames) < 1 {
		log.Printf("server host: error: no hostname")
		return
	}
	log.Printf("server host: %s", hostnames[0])
}

// GenHostname generates the hostname for a service given the hostname_suffix
// flag.
func GenHostname(hostnameSuffix string) string {
	hostname, err := os.Hostname()
	if err != nil {
		hostname = "generated-" + strconv.Itoa(rand.Intn(1000))
	}
	if hostnameSuffix != "" {
		return hostname + "_" + hostnameSuffix
	}
	return hostname
}

// ValidateMembership performs an rpc to the account service to validate a
// user's membership.
func ValidateMembership(inCtx context.Context, accountClient accountpb.AccountClient) (premium bool, name, token, requestedMembership string, err error) {
	// Pull metadata from incoming context.
	inMd, ok := metadata.FromIncomingContext(inCtx)
	if !ok {
		return false, "", "", "", status.Error(codes.Unauthenticated, "missing user authentication")
	}

	// Pull token and membership from incoming metadata.
	tokens, ok := inMd["authorization"]
	if !ok || len(tokens) < 1 {
		return false, "", "", "", status.Error(codes.Unauthenticated, "missing user authentication")
	}
	memberships, ok := inMd["membership"]
	if !ok || len(memberships) < 1 {
		return false, "", "", "", status.Error(codes.Unauthenticated, "missing user authentication")
	}
	token = tokens[0]
	requestedMembership = memberships[0]
	if requestedMembership != "normal" && requestedMembership != "premium" {
		return false, "", "", "", status.Error(codes.Unauthenticated, "unknown user membership")
	}
	routes, ok := inMd["route"]
	if ok {
	  inCtx = metadata.NewOutgoingContext(inCtx, metadata.Pairs("route", routes[0]))
	}

	// Perform RPC to account client.
	var header metadata.MD
	r, err := accountClient.GetUserInfo(inCtx, &accountpb.GetUserInfoRequest{Token: token}, grpc.Header(&header))
	if err != nil {
		return false, "", "", "", status.Errorf(status.Code(err), "could not get balance: %v", err)
	}
	PrintHostname(header)

	// Parse membership from request response.
	actualMembership := ""
	switch r.GetMembership() {
	case accountpb.MembershipType_PREMIUM:
		actualMembership = "premium"
	case accountpb.MembershipType_NORMAL:
		actualMembership = "normal"
	default:
		return false, "", "", "", status.Error(codes.Unauthenticated, "unrecognized user membership type")
	}

	// Print results & return.
	name = r.GetName()
	success := requestedMembership <= actualMembership
	log.Printf("token: %s, name: %s, membership: %s, requested membership: %s, authentication success: %v", token, name, actualMembership, requestedMembership, success)
	if !success {
		return false, "", "", "", status.Error(codes.Unauthenticated, "requested membership higher than actual membership")
	}
	return requestedMembership != "normal", name, token, requestedMembership, nil
}

// ParseCredentialsType parses the value set for the command line flag -creds.
// Supported values for this flag include 'insecure' and 'xds'. When a nil error
// is returned, the first return value indicates whether xDS credentials are to
// be used. A non-nil error is returned when credsType is set to an unsupported
// credentials type.
func ParseCredentialsType(credsType string) (bool, error) {
	switch credsType {
	case "insecure":
		return false, nil
	case "xds":
		return true, nil
	default:
		return false, fmt.Errorf("-creds set to unsupported value %q. Only supported values are 'insecure' and 'xds'", credsType)
	}
}

// StartAdminServices exposes admin services (csds, health and reflection) on
// the on admin_port. It is a blocking call and will return only when the gRPC
// server exposing the admin services is stopped.
func StartAdminServices(adminPort string) error {
	adminListener, err := net.Listen("tcp", ":"+adminPort)
	if err != nil {
		return fmt.Errorf("failed to listen: %v", err)
	}
	s := grpc.NewServer()
	cleanup, err := admin.Register(s)
	if err != nil {
		return fmt.Errorf("failed to register admin service: %v", err)
	}
	defer cleanup()
	reflection.Register(s)
	healthServer := health.NewServer()
	healthServer.SetServingStatus("", healthpb.HealthCheckResponse_SERVING)
	healthpb.RegisterHealthServer(s, healthServer)
	if err := s.Serve(adminListener); err != nil {
		return fmt.Errorf("failed to serve admin services: %v", err)
	}
	return nil
}
